digdag-core/src/main/java/io/digdag/core/database/DatabaseTaskQueueServer.java

```diff
@@ -0,0 +1,494 @@
+        }
+    }
+
+    private long enqueue(Integer siteId, Integer queueId,
+            int priority, Long uniqueTaskId,
+            byte[] data)
+        throws ResourceConflictException
+    {
+        long id = transaction((handle, dao, ts) -> {
+            long queuedTaskId = catchConflict(() ->
+                dao.insertQueuedTask(siteId, queueId, uniqueTaskId, data),
+                "lock of task id=%d in site id = %d and queue id=%d", uniqueTaskId, siteId, queueId);
+            dao.insertQueuedTaskLock(queuedTaskId, siteId, queueId, priority);
+            return queuedTaskId;
+        }, ResourceConflictException.class);
+
+        noticeEnqueue();
+
+        return id;
+    }
+
+    private byte[] encodeTaskObject(TaskRequest request)
+    {
+        try {
+            return taskObjectMapper.writeValueAsBytes(request);
+        }
+        catch (IOException ex) {
+            throw Throwables.propagate(ex);
+        }
+    }
+
+    private TaskRequest decodeLockedTaskObject(byte[] data, String lockId)
+    {
+        try {
+            return TaskRequest.withLockId(
+                taskObjectMapper.readValue(data, TaskRequest.class),
+                lockId);
+        }
+        catch (IOException ex) {
+            throw Throwables.propagate(ex);
+        }
+    }
+
+    private static String formatSharedTaskLockId(long taskLockId)
+    {
+        return "s" + Long.toString(taskLockId);
+    }
+
+    private static String formatQueueBoundTaskLockId(long taskLockId, int queueId)
+    {
+        return "q" + Long.toString(taskLockId) + "." + Integer.toString(queueId);
+    }
+
+    private static boolean isSharedTaskLockId(String formatted)
+    {
+        return formatted.startsWith("s");
+    }
+
+    private static long parseTaskLockId(String formatted)
+    {
+        return Long.parseLong(formatted.split("\\.", 2)[0].substring(1));
+    }
+
+    private static int parseQueueId(String formatted)
+    {
+        String[] fragments = formatted.split("\\.");
+        if (fragments.length != 2) {
+            throw new IllegalArgumentException("Invalid queue-bound task lock id: " + formatted);
+        }
+        return Integer.parseInt(fragments[1]);
+    }
+
+    @Override
+    public void deleteTask(int siteId, String lockId, String agentId)
+        throws TaskNotFoundException, TaskConflictException
+    {
+        long taskLockId = parseTaskLockId(lockId);
+        deleteTask0(siteId, taskLockId, agentId);
+    }
+
+    private void deleteTask0(int siteId, long taskLockId, String agentId)
+        throws TaskNotFoundException, TaskConflictException
+    {
+        this.<Boolean, TaskNotFoundException, TaskConflictException>transaction((handle, dao, ts) -> {
+            int count;
+
+            count = dao.deleteQueuedTask(siteId, taskLockId);
+            if (count == 0) {
+                throw new TaskNotFoundException("Deleting lock does not exist: lock id=" + taskLockId + " site id=" + siteId);
```

Comment:
I see. About inconsistency,

* If `queued_task_locks` doesn't exist but `queued_tasks` exists, the task won't be dequeued for ever. It is a problem in terms of garbage but not a big deal because it doesn't affect other logics.
* If `queued_task_locks` exists but `queued_tasks` doesn't exist, `lockSharedAgentTasks` will go into `// queued_task is deleted after tryLockSharedAgentTasks call.` section. This is not good because it consumes a garbage slot from max_concurrency...

But deleteTask won't be called anyways if this inconsistency happens. So, force-deleting records here doesn't solve this issue...
I think what we can is using transaction appropriately not to cause this issue.